### ISSUE TITLE
Add support for streams providing the elements of a power set.

### ISSUE THREAD

**Issue by siordache** [beryx/streamplify#1]
This is an _up-for-grabs_ issue, so hurry up and grab it!

Your task is to add support for streams providing the elements of a [power set](https://en.wikipedia.org/wiki/Power_set).

Consult the implementation of the other combinatorics classes (permutations, combinations, Cartesian product) and structure your code in a similar way:
- put your classes in a package named `org.beryx.streamplify.powerset`.
- create a subclass of [LongIndexedSpliterator](../blob/master/streamplify/src/main/java/org/beryx/streamplify/LongIndexedSpliterator.java)  called `LongPowerSet`, which can be used for power sets with a cardinality that fits in a long (that is, for sets with at most 63 elements).
  _Hint_: use [LongPermutations.java](../blob/master/streamplify/src/main/java/org/beryx/streamplify/permutation/LongPermutations.java) as template for this class.
- create a subclass of [BigIntegerIndexedSpliterator](../blob/master/streamplify/src/main/java/org/beryx/streamplify/BigIntegerIndexedSpliterator.java) called `BigIntegerPowerSet`, which allows power sets of virtually any cardinality.
  _Hint_: use [BigIntegerPermutations.java](../blob/master/streamplify/src/main/java/org/beryx/streamplify/permutation/BigIntegerPermutations.java) as template for this class.
- create a [StreamableProxy](../blob/master/streamplify/src/main/java/org/beryx/streamplify/StreamableProxy.java)  called`PowerSet`, which delegates to either _LongPowerSet_ or _BigIntegerPowerSet_.
  _Hint_: use [Permutations.java](../blob/master/streamplify/src/main/java/org/beryx/streamplify/permutation/Permutations.java) as template for this class.
- create an abstract class called `PowerSetSupplier`, which implements [IntArraySupplier](../blob/master/streamplify/src/main/java/org/beryx/streamplify/IntArraySupplier.java) and has two concrete subclasses: the static nested classes `Long` and `BigInteger`.
  _Hint_: use [PermutationSupplier.java](../blob/master/streamplify/src/main/java/org/beryx/streamplify/permutation/PermutationSupplier.java) as template for this class.
  (This is the most challenging class to be implemented, because you have to come up with efficient algorithms for the methods _computeNext()_ and _unrank()_.)

Your pull request should also contain:
- [Spock](http://spockframework.org/) unit tests - see for example [PermutationsSpec.groovy](../blob/master/streamplify/src/test/groovy/org/beryx/streamplify/PermutationsSpec.groovy)
- examples - should be placed in [streamplify-examples](../blob/master/streamplify-examples/src/main/java/org/beryx/streamplify/example)
- documentation (using [Asciidoctor](http://asciidoctor.org/)) - in [user_guide.adoc](../blob/master/doc/user_guide.adoc), describe how to use your new feature

In order to prevent that two people work on the same thing, leave a comment here to claim this issue before starting to work on it.

Happy coding, and don't hesitate to ask for clarification and guidance.

**Comment by brightcoder** [comment]
Hi Siordache,
Can I take up this task ? I would be happy to work on this as my first open source contribution.

**Comment by siordache** [comment]
@brightcoder: It's yours. Have fun!

**Comment by brightcoder** [comment]
@siordache, I am building this project in eclipse gradle but failing to import this project.
Here is the error shown by eclipse: https://paste.kde.org/pliiz0nxz

**Comment by siordache** [comment]
@brightcoder: I fixed the build, so it should work now.

To start working on this pull request:
- fork the repository (click **Fork** in the top-right corner of the project page)
- clone the fork:

```
    git clone https://github.com/brightcoder/streamplify.git
```
- create a branch for your pull request:

```
    cd streamplify
    git checkout -b powerset
```
- open Eclipse
- install the [Groovy plugin](https://github.com/groovy/groovy-eclipse/wiki).
  Update site: http://dist.springsource.org/snapshot/GRECLIPSE/e4.6/
- install Buildship 1.0.21 (the import does not work with older versions).
  Update site: http://download.eclipse.org/buildship/updates/e46/releases/1.0/1.0.21.v20161010-1640/
- import the project using the _Gradle Project_ wizard

**Comment by brightcoder** [comment]
Thanks. This worked.

**Comment by brightcoder** [comment]
In PowerSetSupplier, we will require two fields. 
int length and int[] currentSet.
The size of currentSet will be dynamic unlike in PermutationSupplier.java.
For example, length=3
index=0, currentSet = {}
index=1, currentSet = {0}
index=2, currentSet = {1}
index=3, currentSet = {0,1}
index=4, currentSet = {2}
index=5, currentSet = {0,2}
index=6, currentSet = {1,2}
index=7, currentSet = {0,1,2}
@siordache, let me know if you see any better alternative to this.

**Comment by siordache** [comment]
@brightcoder That's absolutely ok.

**Comment by brightcoder** [comment]
Default implementation of getNextSequence in IntArraySupplier.java will not work for this case because it assumes the same reference for current and next sequence. But computeNext() method for this case will have to create new array with new size each time.

**Comment by siordache** [comment]
@brightcoder: You're right. Your PowerSetSupplier has to override the default implementation of getNextSequence().

It's difficult to predict whether creating a new array on each call of computeNext() will cause a significant performance loss. Let's ignore this issue for now. Premature optimization is the root of all evil 😃

**Comment by brightcoder** [comment]
I have coded the solution with the following approach. If this looks good to all of us then I can start writing the test cases.
We maintain a binary counter in form of array and increment by 1 in computeNext(). Occurrence of 1's in binary counter corresponds to inclusion of corresponding element.
For unrank(), compute binary representation of the current index on the fly and include elements corresponding to 1's.

I have one doubt:
I am finding it difficult to understand the role of shuffler. Why do we supply shuffledIndex instead of index to the valueSupplier.apply in tryAdvance of LongIndexedSpliterator.java ?

**Comment by siordache** [comment]
@brightcoder Yes, this is a good approach.

Shuffling a stream implies defining a permutation of its elements. Because our elements are indexed, we can instead define a permutation of the indices. This is the job of the shuffler. If the stream should be shuffled, tryAdvance() retrieves the element corresponding to the shuffledIndex instead of the original index. This way, the elements are provided in a permuted order.

For a non-shuffled stream the indices have consecutive values. Therefore, the valueSupplier calls most of the time computeNext() in its getNextSequence() method. For a shuffled stream, the indices have no longer consecutive values and the valueSupplier will have to call most of the time the unrank() method, which is usually more expensive. This means that shuffling a stream has a negative impact on performance.

You don't have to understand how the shuffling works, but if you're interested, I blogged about it [here](http://beryx.org/blog/2016-10-16-/shuffling-large-sequences/).

**Comment by siordache** [comment]
Implemented in #5  by @brightcoder.

Will be released in 1.1.0.
